packages/itwin/tree-widget/src/tree-widget-react/components/trees/categories-tree/internal/visibility/CategoriesTreeVisibilityHandler.ts

```diff
@@ -0,0 +1,400 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+import { defer, EMPTY, from, map, merge, mergeMap, of } from "rxjs";
+import { assert, Id64 } from "@itwin/core-bentley";
+import { HierarchyNode } from "@itwin/presentation-hierarchies";
+import { createVisibilityStatus } from "../../../common/internal/Tooltip.js";
+import { HierarchyVisibilityHandlerImpl } from "../../../common/internal/useTreeHooks/UseCachedVisibility.js";
+import { getClassesByView, releaseMainThreadOnItemsCount } from "../../../common/internal/Utils.js";
+import { mergeVisibilityStatuses } from "../../../common/internal/VisibilityUtils.js";
+import { CategoriesTreeNode } from "../CategoriesTreeNode.js";
+import { CategoriesTreeVisibilityHelper } from "./CategoriesTreeVisibilityHelper.js";
+import { createFilteredCategoriesTree } from "./FilteredTree.js";
+
+import type { Observable } from "rxjs";
+import type { Viewport } from "@itwin/core-frontend";
+import type { GroupingHierarchyNode, HierarchyFilteringPath } from "@itwin/presentation-hierarchies";
+import type { ECClassHierarchyInspector } from "@itwin/presentation-shared";
+import type { AlwaysAndNeverDrawnElementInfo } from "../../../common/internal/AlwaysAndNeverDrawnElementInfo.js";
+import type { ElementId, ModelId } from "../../../common/internal/Types.js";
+import type { VisibilityStatus } from "../../../common/UseHierarchyVisibility.js";
+import type { FilteredTree } from "../../../common/internal/visibility/BaseFilteredTree.js";
+import type { BaseIdsCache, TreeSpecificVisibilityHandler } from "../../../common/internal/visibility/BaseVisibilityHelper.js";
+import type { CategoriesTreeIdsCache } from "../CategoriesTreeIdsCache.js";
+import type { CategoriesTreeFilterTargets } from "./FilteredTree.js";
+
+/** @internal */
+export interface CategoriesTreeVisibilityHandlerProps {
+  idsCache: CategoriesTreeIdsCache;
+  viewport: Viewport;
+  alwaysAndNeverDrawnElementInfo: AlwaysAndNeverDrawnElementInfo;
+}
+
+/**
+ * Handles visibility status of categories tree nodes.
+ *
+ * This handler knows how to get and change visibility status of nodes created by hierarchy definition.
+ * @internal
+ */
+export class CategoriesTreeVisibilityHandler implements Disposable, TreeSpecificVisibilityHandler<CategoriesTreeFilterTargets> {
+  private _visibilityHelper: CategoriesTreeVisibilityHelper;
+  private _elementType: "GeometricElement3d" | "GeometricElement2d";
+  private _categoryType: "DrawingCategory" | "SpatialCategory";
+  private _modelType: "GeometricModel3d" | "GeometricModel2d";
+  constructor(private readonly _props: CategoriesTreeVisibilityHandlerProps) {
```

Comment:
Should be use `#props` for private members same as in `CategoriesTreeVisibilityHelper`? Maybe use `#` for other private members?

Reply:
We should be moving towards using `#` for privates. Newly added or heavily refactored classes should be using `#`.